trunk/paradiseo-eo/doc/man/man3/eoIntBounds.3
.TH "eoIntBounds" 3 "19 Oct 2006" "Version 0.9.4-cvs" "EO" \" -*- nroff -*-
.ad l
.nh
.SH NAME
eoIntBounds \- Defines bound classes for real numbers.  

.PP
.SH SYNOPSIS
.br
.PP
\fC#include <es/eoIntBounds.h>\fP
.PP
Inherits \fBeoPersistent\fP.
.PP
Inherited by \fBeoGeneralIntBounds\fP, \fBeoIntAboveBound\fP, \fBeoIntBelowBound\fP, \fBeoIntInterval\fP, and \fBeoIntNoBounds\fP.
.PP
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "virtual bool \fBisBounded\fP (void) const =0"
.br
.RI "\fISelf-Test: true if ***both*** a min and a max. \fP"
.ti -1c
.RI "virtual bool \fBhasNoBoundAtAll\fP (void) const =0"
.br
.RI "\fISelf-Test: true if no min ***and*** no max hence no further need to test/truncate/fold anything. \fP"
.ti -1c
.RI "virtual bool \fBisMinBounded\fP (void) const =0"
.br
.RI "\fISelf-Test: bounded from below??? \fP"
.ti -1c
.RI "virtual bool \fBisMaxBounded\fP (void) const =0"
.br
.RI "\fISelf-Test: bounded from above??? \fP"
.ti -1c
.RI "virtual bool \fBisInBounds\fP (double) const =0"
.br
.RI "\fITest on a value: is it in bounds? \fP"
.ti -1c
.RI "virtual void \fBfoldsInBounds\fP (double &) const =0"
.br
.RI "\fIPut value back into bounds - by folding back and forth. \fP"
.ti -1c
.RI "virtual void \fBfoldsInBounds\fP (long int &i) const "
.br
.RI "\fIfoldsInBounds for ints: call the method for double and convert back \fP"
.ti -1c
.RI "virtual void \fBtruncate\fP (double &) const =0"
.br
.RI "\fIPut value back into bounds - by truncating to a boundary value. \fP"
.ti -1c
.RI "virtual void \fBtruncate\fP (long int &i) const "
.br
.RI "\fItruncate for ints: call the method for double and convert back \fP"
.ti -1c
.RI "virtual long int \fBminimum\fP () const =0"
.br
.RI "\fIget minimum value ::exception if does not exist \fP"
.ti -1c
.RI "virtual long int \fBmaximum\fP () const =0"
.br
.RI "\fIget maximum value ::exception if does not exist \fP"
.ti -1c
.RI "virtual long int \fBrange\fP () const =0"
.br
.RI "\fIget range ::exception if unbounded \fP"
.ti -1c
.RI "virtual double \fBuniform\fP (\fBeoRng\fP &_rng=eo::rng) const =0"
.br
.RI "\fIrandom generator of uniform numbers in bounds uses same naming convention than eo::rng ::exception if unbounded \fP"
.ti -1c
.RI "virtual long int \fBrandom\fP (\fBeoRng\fP &_rng=eo::rng) const =0"
.br
.ti -1c
.RI "virtual \fBeoIntBounds\fP * \fBdup\fP () const =0"
.br
.RI "\fIfor memory managements - ugly \fP"
.in -1c
.SH "Detailed Description"
.PP 
Defines bound classes for real numbers. 

Scalar type: ------------ Basic class is eoIntBounds, a pure virtual.
.PP
The following pure virtual methods are to be used in mutations:
.IP "\(bu" 2
void foldsInBounds(long int &) that folds any value that falls out of the bounds back into the bounds, by bouncing on the limit (if any)
.IP "\(bu" 2
bool isInBounds(long int) that simply says whether or not the argument is in the bounds
.IP "\(bu" 2
void truncate(long int &) that set the argument to the bound value it it exceeds it
.PP
.PP
So mutation can choose
.IP "\(bu" 2
iterate trying until they fall in bounds,
.IP "\(bu" 2
only try once and 'restd::pair' by using the foldsInBounds method
.IP "\(bu" 2
only try once and restd::pair using the truncate method (will create a huge bias toward the bound if the soluiton is not far from the bounds)
.PP
.PP
There is also a \fBuniform()\fP method that generates a uniform value (if possible, i.e. if bounded) in the interval.
.PP
Derived class are \fBeoIntInterval\fP that holds a minimum and maximum value, \fBeoIntNoBounds\fP the 'unbounded bounds' (-infinity, +infinity) \fBeoIntBelowBound\fP the half-bounded interval [min, +infinity) \fBeoIntAboveBound\fP the half-bounded interval (-infinity, max]
.PP
THis file also contains the declaration of *the* global object that is the unbounded bound 
.PP
Definition at line 75 of file eoIntBounds.h.

.SH "Author"
.PP 
Generated automatically by Doxygen for EO from the source code.


eo/NEWS
* release 1.0 (not yet released)
  - Implement CMA-ES.
  - Update introductory pages of documentation and webpage.
  - Add Microsoft Visual C++ project support files (for Visual Studio 2003 and 2005)
  - Upgrade Teamplates/ script to create a complete standalon EO project (using autotools)
  - Remove support for pre-standard C++ compiler (i.e. gcc-2.x), which allows to
    clean up the code considerably. 
    Assume availability of sstream and limits.

* release 0.9.3z.1 (1. Oct. 2005)
  - Support gcc-3.4 and gcc.4.x.
  - Provide full automake/autoconf/configure support.


# Local Variables:
# coding: iso-8859-1
# mode: text
# fill-column: 80
# End:


eo/src/utils/eoIntBounds.cpp
#ifdef _MSC_VER
// to avoid long name warnings
#pragma warning(disable:4786)
#endif

#ifdef HAVE_CONFIG_H
#include <config.h>
#endif

#include <ctime>
#include <sstream>

#include "eoIntBounds.h"


// the global dummy bounds
// (used for unbounded variables when bounds are required)
eoIntNoBounds eoDummyIntNoBounds;

///////////// helper read functions defined in eoRealBounds.cpp
extern bool remove_leading(std::string & _s, const std::string _delim);
extern double read_double(std::string _s);
extern long int read_int(std::string _s);


/** the constructor for eoGeneralIntBound - from a string
 */
eoIntBounds* eoGeneralIntBounds::getBoundsFromString(std::string _value)
{
  // now read
  std::string delim(",; ");
  std::string beginOrClose("[(])");
  if (!remove_leading(_value, delim)) // only delimiters were left
    throw std::runtime_error("Syntax error in eoGeneralIntBounds Ctor");

  // look for opening char
  size_t posDeb = _value.find_first_of(beginOrClose);	// allow ]a,b]
  if (posDeb >= _value.size())	// nothing left to read
    throw std::runtime_error("Syntax error in eoGeneralIntBounds Ctor");

  // ending char: next {}() after posDeb
  size_t posFin = _value.find_first_of(beginOrClose,posDeb+1);
  if (posFin >= _value.size())	// not found
    throw std::runtime_error("Syntax error in eoGeneralIntBounds Ctor");

  // the bounds
  std::string sBounds = _value.substr(posDeb+1, posFin-posDeb-1);
  // and remove from original string
  _value = _value.substr(posFin+1);

  remove_leading(sBounds, delim);
  size_t posDelim = sBounds.find_first_of(delim);
  if (posDelim >= sBounds.size())
    throw std::runtime_error("Syntax error in eoGeneralIntBounds Ctor");

      bool minBounded=false, maxBounded=false;
      long int minBound=0, maxBound=0;

      // min bound
      std::string sMinBounds = sBounds.substr(0,posDelim);

      if ( (sMinBounds != std::string("-inf")) &&
	   (sMinBounds != std::string("-infinity"))
	   )
	{
	  minBounded = true;
	  minBound = read_int(sMinBounds);
	}

      // max bound
      size_t posEndDelim = sBounds.find_first_not_of(delim,posDelim);

      std::string sMaxBounds = sBounds.substr(posEndDelim);

      if ( (sMaxBounds != std::string("+inf")) &&
	   (sMaxBounds != std::string("+infinity"))
	   )
	{
	  maxBounded = true;
	  maxBound = read_int(sMaxBounds);
	}

      // now create the embedded eoIntBounds object
      eoIntBounds * locBound;
      if (minBounded && maxBounded)
	{
	  if (maxBound <= minBound)
	    throw std::runtime_error("Syntax error in eoGeneralIntBounds Ctor");
	  locBound = new eoIntInterval(minBound, maxBound);
	}
      else if (!minBounded && !maxBounded)	// no bound at all
	locBound = new eoIntNoBounds;
      else if (!minBounded && maxBounded)
	locBound = new eoIntAboveBound(maxBound);
      else if (minBounded && !maxBounded)
	locBound = new eoIntBelowBound(minBound);
      return locBound;
}
